Remove feature guard on lifetime dependence inference

meg-gupta opened the pull request:
Lifetime dependence inference runs on all function results that are ~Escapable. It was guarded by the NonEscapable feature flag. This PR removes the feature guard. This should theoretically be an NFC because ~Escapable types are flagged by Sema without the feature flag.

meg-gupta committed:
[NFC] Rename helper function for @_noImplicitCopy and @_eagerMove

meg-gupta committed:
[NFC] Add test for ~Escapable features without NonEscapable feature flag

meg-gupta committed:
Add Escapable conformance to some builtin functions

meg-gupta committed:
Update some test configs where stdlib is loaded later

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test macOS platform

meg-gupta commented:
@swift-ci smoke test macOS platform

meg-gupta commented:
@swift-ci smoke test macOS platform

meg-gupta committed:
C++Interop: Don't create default argument expressions that are templated

SwiftDeclSynthesizer::makeDefaultArgument does not create substituion map while
creating the implicit FuncDecl which leads to TypeChecker complaining about
the absence when there are type parameters. Avoid importing such C++ functions for now.

meg-gupta committed:
XFAIL typechecker test which leads to a cycle in the requirement machine

This PR introduces a `mapTypeIntoContext` in `InterfaceTypeRequest::evaluate` which introduces a cycle in this test.

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test

egorzhdan commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -2660,6 +2658,13 @@ static ParamDecl *getParameterInfo(ClangImporter::Implementation *impl,
       !isa<clang::CXXConstructorDecl>(param->getDeclContext()) &&
       impl->isDefaultArgSafeToImport(param)) {
+
+    // If this was a default argument expression with template parameter, return
+    // std::nullopt
+    if (isTemplated) {
+      return std::nullopt;
+    }
```
This makes the entire function not available in Swift, but I don't think that's what we really want (and that would be a source compatibility breakage). I think we should just drop the default expression for this parameter by adding `&& !isTemplated` to the if-condition above.

meg-gupta replied:
Done

egorzhdan commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -2616,12 +2616,10 @@ bool ClangImporter::Implementation::isDefaultArgSafeToImport(
+static std::optional<ParamDecl *> getParameterInfo(
+    ClangImporter::Implementation *impl, const clang::ParmVarDecl *param,
+    const Identifier &name, const swift::Type &swiftParamTy, const bool isInOut,
+    const bool isParamTypeImplicitlyUnwrapped, const bool isTemplated) {
```
Could we check for `isa<clang::FunctionTemplateDecl>(param->getDeclContext())` instead of having an extra `isTemplated` parameter?

meg-gupta replied:
Done.

egorzhdan reviewed (changes requested):
I only have a few comments on the changes to C++ interop.

egorzhdan reviewed (approved):
The C++ interop changes LGTM now, thanks!
I haven't reviewed the rest of the PR.

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test

meg-gupta committed:
Prespecialize standard library only if it is loaded

We currently load it for prespecialization when it wasn't loaded initially.

This causes an inadvertant issue for invertible protocols.

When we don’t have the stdlib loaded initially, we “synthesize” the
invertible protocol from the Builtin module by creating a new `ProtocolDecl*`
and stashing it on the `ASTContext`.
If the stdlib gets loaded later, deserialized stdlib types conform to the deserialized `Escapable` protocol
which has a different `ProtocolDecl *` pointer for `Escapable`.
So queries like `conformsToInvertible` fail because they are using the wrong `ProtocolDecl*`
for `Copyable`/`Escapable` while looking up the ConformanceTable.

meg-gupta committed:
Remove NonescapableTypes feature guard while querying lifetime dependence

atrick reviewed (approved):
LGTM

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test
